### ISSUE TITLE
Reason for not being optimized

### ISSUE THREAD

**Issue by Frank207507** [cortwave/MosaicSDF#1]
Thank you for your great work！
 I ran your optimization code and found that the reason for the failure of optimization was that loss did not have backpropagation, but only needed just add a loss.backward() before self.optimizer.step()

**Comment by cortwave** [comment]
Wow, thank you so much! Will need to try it again, such a stupid mistake :sweat_smile:

**Comment by cortwave** [comment]
@Jingwei-Jiang I pushed a fix for optimization. Also, I changed the sampling approach and fixed bugs with marching cubes and SDF gt values during optimization (SDF library I use returns positive values inside the mesh and negative outside). Now optimization converges but I still see no difference between optimized and not-optimized versions (maybe I need to try it on different meshes, not only octopus, because octopus now looks the same as gt even without optimization). But due to the marching cube fix reconstructed meshes look much better now

**Comment by Frank207507** [comment]
Significant improvement! 
 I ran a few other 3D objects, and they were all fairly similar. Additionally, I evaluated the chamfer distance before and after optimization (after normalizing the mesh) and found a slight improvement, from 0.05 to 0.048. 
By the way, pcu has implemented the calculation of chamfer distance, where 
`average_distance = pcu.chamfer_distance(sample_points_1, sample_points_2)`.
I think that there's more to explore in terms of the loss function.

![截图 2024-04-26 20-41-46](https://github.com/cortwave/MosaicSDF/assets/82004964/d063893c-e9d7-48b4-be43-b560bdbf894a)
